guides/docs/testing/testing_controllers.md

```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 
 Running the test tells us we need a `render/2` function that can pattern match on `"show.json"`:
 
 ```elixir
 defmodule HelloWeb.UserView do
   use HelloWeb, :view
 
-  def render("index.json", %{users: users}) do
-    %{data: render_many(users, HelloWeb.UserView, "user.json")}
+  def render("index.json", %{data: users}) do
+    %{data:
+      render_many( users, HelloWeb.UserView, "user.json", as: :data)
+      }
   end
 
-  def render("show.json", %{user: user}) do
-    %{data: render_one(user, HelloWeb.UserView, "user.json")}
+  def render("show.json", %{data: user}) do
+    %{data:
+      render_one( user, HelloWeb.UserView, "user.json", as: :data)
+      }
   end
 
-  def render("user.json", %{user: user}) do
-    %{name: user.name, email: user.email}
+  def render("user.json", %{data: user}) do
+    %{
+      name: user.name,
+      email: user.email
+      # skipping password, inserted_at, and updated_at
+    }
   end
-
 end
 ```
 
+Notice the "show.json" rendering path uses `render_one/4` instead of
+`render_many/4` because it is only rendering a single user, not a list.
+
+
 When we run the test again, it passes.
 
+### No show
+
 The last item we'll cover is the case where we don't find a user in `show/2`.
 
 Try this one on your own and see what you come up with. One possible solution will be given below.
 
-Walking through our TDD steps, we add a test that supplies a non existent user id to `user_path` which returns a 404 status and an error message:
+Walking through our TDD steps, we add a test that supplies a
+non-existent user id to `user_path` which returns a 404 status and an
+error message.  One interesting problem here is how we might define a
+"non-existent" id.  We could just pick a large integer, but who's to
+say some future test won't generate thousands of test users and break
+our test?  Instead of going bigger, we can also go the other way.
+Database ids tend to start at 1 and increase forever.  Negative
+numbers are perfectly valid integers, and yet never used for database
+ids.  So we'll pick -1 as our "unobtainable" user id, which *should*
+always fail.
+
 
 ```elixir
-test "Responds with a message indicating user not found" do
-  response = build_conn()
-  |> get(user_path(build_conn(), :show, 300))
+# test/hello_web/controllers/user_controller_test.exs line 55 (or so)
+
+test "Responds with a message indicating user not found", %{conn:  conn} do
+  response = conn
+  |> get(user_path(conn, :show, -1 ))
   |> json_response(404)
 
-  expected = %{ "error" => "User not found." }
+  expected = %{
+    "errors" => "Resource not found"
+  }
 
   assert response == expected
 end
 ```
 
 We want a HTTP code of 404 to notify the requester that this resource was not found, as well as an accompanying error message.
 
-Our controller action:
+Our controller action now needs to handle the error thrown by Ecto:
```

Comment:
I don't think we should do this. The previous version should work without explicitly using a try/catch. If you wanted to match on the error, I'd recommend not using `Accounts.get_user!`, but defining an `Accounts.get_user` instead.

Reply:
I don't really care much, but my thinking was the opposite of yours!

I didn't want to get into defining Accounts.get_user (something lots of guides do).  I chose instead to discuss catching exceptions, which I haven't seen much of in my reading of Phoenix guides. 

Thinking more now, I would argue (again, weakly, not my project I'll defer to your judgement) to keep the exception handling in.  The code generated by the phx.gen.json generator looks like this:

```
  def show(conn, %{"id" => id}) do
    user = Accounts.get_user!(id)
    render(conn, "show.json", user: user)
  end
```

If one passes an invalid id to that, it will throw.  The auto-generated tests do NOT test that case (I just double checked).  

I'm not sure why the code generators were designed with the get_THING!(id) semantics that advertise that they will throw a not found error, but if it throws then it has to be handled properly.  If everybody resorts to re-implementing get_THING! into get_THING (that does not throw), then why is the throwing version in the automatic generator in the first place?

Reply:
When there is an `Ecto.NoResultsFound`, the `Plug.Exception` protocol is used to render the error. https://github.com/phoenixframework/phoenix_ecto/blob/master/lib/phoenix_ecto/plug.ex

I think the discussion around this worth opening another issue for though. So let's leave it for now and open a new issue.

Reply:
fixed.